perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -1455,6 +1469,107 @@ private void categorizeMembersForSynchronization(PerunSession sess, List<RichMem
+		}
+
+		//get all logins of this extSource from the group and also prepare all other members in group without this extSource
+		List<RichMember> richMembersWithoutLogin = new ArrayList<>();
+		Map<String, RichMember> groupLogins = new HashMap<>();
+		for(RichMember rm: groupMembers) {
+			List<UserExtSource> rmues = rm.getUserExtSources();
+			boolean found = false;
+			for(UserExtSource ues: rmues) {
+				if(ues.getExtSource().equals(membersSource)) {
+					groupLogins.put(ues.getLogin(), rm);
+					found = true;
+					break;
+				}
+			}
+			if(!found) richMembersWithoutLogin.add(rm);
+		}
+
+		//prepare special structures of members in group and in source (members in group and source are not interesting for lightweight synchronization)
+		List<RichMember> existInGroupNotInSource = new ArrayList(richMembersWithoutLogin);
+		Map<String, Map<String, String>> existInSourceNotInGroup = new HashMap<>();
+		//found all richMembers who are new from ext source (not in group yet)
+		for(String loginInSource: sourceLogins.keySet()) {
+			if(!groupLogins.containsKey(loginInSource)) {
+				existInSourceNotInGroup.put(loginInSource, sourceLogins.get(loginInSource));
+				groupLogins.remove(loginInSource);
+			} // skip in other case (updated users are not useful for lightweight synchronization
+		}
+		//for the rest of members in group, add them to the existing members in group, not in source
+		for(String loginInGroup: groupLogins.keySet()) {
+			existInGroupNotInSource.add(groupLogins.get(loginInGroup));
+		}
+
+		//for missing members get all information
+		List<Map<String, String>> subjectsForCandidates = new ArrayList<>();
+		for(String login: existInSourceNotInGroup.keySet()) {
+			subjectsForCandidates.add(existInSourceNotInGroup.get(login));
+		}
+		List<Candidate> subjectsCandidates = convertSubjectsToCandidates(sess, subjectsForCandidates, membersSource, source, skippedMembers);
+
+		//mapping structure for more efficient searching in members' userExtSources
+		Map<UserExtSource, RichMember> mappingStructure = new HashMap<>();
+		for(RichMember rm: existInGroupNotInSource) {
+			for(UserExtSource ues: rm.getUserExtSources()) {
+				mappingStructure.put(ues, rm);
+			}
+		}
+
+		//some of members in group and not in source will be removed so add them to membersToRemove list
+		membersToRemove = existInGroupNotInSource;
+		//try to find richMembers in group who has his own existence also in extSource and remove them from membersToRemove
+		for(Candidate candidate: subjectsCandidates) {
+			List<UserExtSource> candidateExtSources = candidate.getUserExtSources();
+			boolean found = false;
+			for(UserExtSource key: candidateExtSources) {
+				//one of candidates in extSource is our member in this group, remove it from the list of members to remove
+				if(mappingStructure.containsKey(key)) {
+					found = true;
+					membersToRemove.remove(mappingStructure.get(key));
+					//skip other sources of this candidate
+					break;
```

Comment:
It's possible to map "candidateExtSources" to two or more members from "membersToRemove". Removing this "break" might help. Of course, it will be better to detect such situation and at least log it.